bbc/simorgh
# WSTEAM1-440 - Next.JS page error handling

Resolves [JIRA WSTEAM1-440](https://jira.dev.bbc.co.uk/browse/WSTEAM1-440)

Overall changes
======
Renders the correct error page for the service based on the `status` code returned from the server side fetch.

Code changes
======
- Removes the `applyBasicPageHandlers` function from the Live page 💥 
- Adds the contexts and `PageWrapper` component to the new custom `_app.tsx` file, so that all Next.JS pages will get these components added to the render
- Conditionally checks the `status` of the response to determine whether or not to render the `ErrorPage` component. We can't unfortunately use the built-in Custom Errors https://nextjs.org/docs/pages/building-your-application/routing/custom-error, as we need to personalise the error pages based on the service. Next.JS doesn't allow you to pass props to the error pages using their built-in error page mechanism, which we need to do to get the translations for the Error page. May be worth looking into Static Generation for these pages, either from us or upstream
- Creates a new `MvtExperiment` type is this will be used across apps
- Creates new `FetchError` type to type check `try/catch` fetch errors
- Adds a generic route to catch-all routes not yet supported in the Next.JS. Not sure if this is really needed, more of a development aid than anything as the other routes will be handled by the Express app and upstream services preventing undefined routes from being reached

Testing
======
1. cd to the `ws-nextjs-app` directory and run `yarn dev`
2. Visit http://localhost:7081/kyrgyz/new_live/cy3p57772znt - confirm the page renders
3. Visit http://localhost:7081/kyrgyz/new_live/cy3p57772zst - confirm the 404 page renders
4. Visit http://localhost:7081/kyrgyz/somerandompage - confirm the 404 page renders
5. 500 testing is more tricky. You can test locally by editing this line: https://github.com/bbc/simorgh/blob/bd81c6bfd5e65bd554df71183b8864a4de5de505/ws-nextjs-app/pages/%5Bservice%5D/new_live/%5Bid%5D/%5B%5B...variant%5D%5D.page.tsx#L138 and changing it to `status: 500` and saving and refreshing the page

Helpful Links
======
_Add Links to useful resources related to this PR if applicable._

[Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)

[Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)
